rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+3. Create datasets and iterators on workers.
+4. Create the replica function that takes an iterator as input, trace it and register it on all workers.  Note: a function may create variables as well. If not specified, they will be created on parameter servers as well.
+5. Dispatch the step function on one available worker.
+6. Repeat 5 until the end of epoch.
+7. Repeat 5 - 6 until the stop criteria is reached.
+
+
+### Interfaces
+
+One of our goals is to make `ParameterServerStrategy`’s API consistent with other strategies so that users can easily switch between them. This may be challenging due to the fundamental difference between synchronous and asynchronous training. Therefore, we try to use most of the Distribution Strategy’ APIs but occasionally we will create APIs that still make sense in other strategies as well.
+
+**Note: all proposed APIs in this document are tentative and only valid for our first version. We will revisit them once we get enough feedback from users and the community.**
+
+
+#### Constraints
+
+Function is first-class citizen. Users should only schedule functions instead of running individual ops, in addition to creating variables.
+
+Users can occasionally run individual ops on the client, only for reporting purposes such as printing a metric’s value.
+
+
+#### Schedule/Join Primitives
+
+The `strategy.run` API was initially developed for synchronous training. We propose a new pair of primitives to
+
+
+*   hide the details of load-balancing, fault tolerance and dynamic scheduling
+*   expose the non-blocking semantics to users.
+
+To enable scheduling a function on any worker, we recommend users create the same dataset, but may shuffle differently, on all workers via `strategy.experimental_distribute_datasets_from_function` API.
+
+
+```python
+class ParameterServerStrategyV2:
+  
+  def schedule(self, replica_fn, args=(), kwargs=()):
+    """Schedule the `replica_fn` on all replicas in a sync group (a worker).
+   
+    Schedule the `replica_fn` on all replicas in a sync group that is available,    
+    returns a future of PerReplica immediately if `function` has return values.
+
+    It implements at-least-once semantics for function execution. If a worker
+    fails, it will try to reschedule the function on another replica group or throw     
+    an exception to users. So this method assumes that function execution can be 
+    out of order and function inputs are shared between sync groups. 
+
+    We don't support the cases where `args` or `kwargs` are bound to a specific
+    sync group. We will consider supporting them in the future.
+
+    If there are barriers in `replica_fn`, it is users' responsibility to make
+    sure they won't cause deadlock.
+    """
+    pass
+  
+  def join(self):
+    """Wait until all scheduled functions are finished.
+
+    Raises an error if any of the functions fails to execute. In this case,
+    there is no guarantee that non-failing functions will complete.
```

Comment:
Based on recent discussion, we plan to revise this so that if both 1) no functions are left in the queue to be scheduled, 2) no function is being executed on remote workers, are satisfied, join() returns. Before that happens, it blocks (for now without a timeout). 

Also, we may design it in a way that if a worker fails to execute the function, the function goes back to the queue for other workers to execute. With this, there would not be the complication of user needing to handle an error (since the user may not know what to do, other than retrying, anyway).

Reply:
For the failure question, the client only distinguishes between retryable and non-retryable errors. For retryable errors such as worker preemption, the client will try to find a new worker to run. For non-retryable errors, the client will report it to users upon `join()`ing or upon materializing the function return value.